// reason-react/React.js
// Generated by Melange


const Synthetic = {};

const Clipboard = {};

const Composition = {};

const Keyboard = {};

const Focus = {};

const Form = {};

const Mouse = {};

const Drag = {};

const Pointer = {};

const $$Selection = {};

const $$Touch = {};

const UI = {};

const Wheel = {};

const Media = {};

const $$Image = {};

const $$Animation = {};

const Transition = {};

const Ref = {};

const Children = {};

const Context = {};

const Fragment = {};

const StrictMode = {};

const Suspense = {};

const Uncurried = {};

const Experimental = {};

const $$Event = {
  Synthetic: Synthetic,
  Clipboard: Clipboard,
  Composition: Composition,
  Drag: Drag,
  Keyboard: Keyboard,
  Focus: Focus,
  Form: Form,
  Mouse: Mouse,
  Pointer: Pointer,
  $$Selection: $$Selection,
  $$Touch: $$Touch,
  UI: UI,
  Wheel: Wheel,
  Media: Media,
  $$Image: $$Image,
  $$Animation: $$Animation,
  Transition: Transition
};

export {
  Ref ,
  Children ,
  Context ,
  Fragment ,
  StrictMode ,
  Suspense ,
  Uncurried ,
  Experimental ,
  $$Event ,
}
/* No side effect */
