// melange/queue.js
// Generated by Melange

import * as Caml_exceptions from "melange.js/caml_exceptions.js";
import * as Caml_js_exceptions from "melange.js/caml_js_exceptions.js";
import * as Caml_option from "melange.js/caml_option.js";
import * as Curry from "melange.js/curry.js";
import * as Stdlib__Seq from "./seq.js";

const Empty = /* @__PURE__ */Caml_exceptions.create("Stdlib.Queue.Empty");

function create(param) {
  return {
          length: 0,
          first: /* Nil */0,
          last: /* Nil */0
        };
}

function clear(q) {
  q.length = 0;
  q.first = /* Nil */0;
  q.last = /* Nil */0;
}

function add(x, q) {
  const cell = /* Cons */{
    content: x,
    next: /* Nil */0
  };
  const last = q.last;
  if (last) {
    q.length = q.length + 1 | 0;
    last.next = cell;
    q.last = cell;
  } else {
    q.length = 1;
    q.first = cell;
    q.last = cell;
  }
}

function peek(q) {
  const match = q.first;
  if (match) {
    return match.content;
  }
  throw new Caml_js_exceptions.MelangeError(Empty, {
            MEL_EXN_ID: Empty
          });
}

function peek_opt(q) {
  const match = q.first;
  if (match) {
    return Caml_option.some(match.content);
  }
  
}

function take(q) {
  const match = q.first;
  if (match) {
    const content = match.content;
    const match$1 = match.next;
    if (match$1) {
      const next = match.next;
      q.length = q.length - 1 | 0;
      q.first = next;
      return content;
    }
    clear(q);
    return content;
  }
  throw new Caml_js_exceptions.MelangeError(Empty, {
            MEL_EXN_ID: Empty
          });
}

function take_opt(q) {
  const match = q.first;
  if (!match) {
    return ;
  }
  const content = match.content;
  const match$1 = match.next;
  if (match$1) {
    const next = match.next;
    q.length = q.length - 1 | 0;
    q.first = next;
    return Caml_option.some(content);
  }
  clear(q);
  return Caml_option.some(content);
}

function copy(q) {
  let q_res = {
    length: q.length,
    first: /* Nil */0,
    last: /* Nil */0
  };
  let _prev = /* Nil */0;
  let _cell = q.first;
  while(true) {
    const cell = _cell;
    const prev = _prev;
    if (cell) {
      const next = cell.next;
      const res = /* Cons */{
        content: cell.content,
        next: /* Nil */0
      };
      if (prev) {
        prev.next = res;
      } else {
        q_res.first = res;
      }
      _cell = next;
      _prev = res;
      continue ;
    }
    q_res.last = prev;
    return q_res;
  };
}

function is_empty(q) {
  return q.length === 0;
}

function length(q) {
  return q.length;
}

function iter(f, q) {
  let _cell = q.first;
  while(true) {
    const cell = _cell;
    if (!cell) {
      return ;
    }
    const next = cell.next;
    Curry._1(f, cell.content);
    _cell = next;
    continue ;
  };
}

function fold(f, accu, q) {
  let _accu = accu;
  let _cell = q.first;
  while(true) {
    const cell = _cell;
    const accu$1 = _accu;
    if (!cell) {
      return accu$1;
    }
    const next = cell.next;
    const accu$2 = Curry._2(f, accu$1, cell.content);
    _cell = next;
    _accu = accu$2;
    continue ;
  };
}

function transfer(q1, q2) {
  if (q1.length <= 0) {
    return ;
  }
  const last = q2.last;
  if (last) {
    q2.length = q2.length + q1.length | 0;
    last.next = q1.first;
    q2.last = q1.last;
    return clear(q1);
  } else {
    q2.length = q1.length;
    q2.first = q1.first;
    q2.last = q1.last;
    return clear(q1);
  }
}

function to_seq(q) {
  const aux = function (c, param) {
    if (!c) {
      return /* Nil */0;
    }
    const next = c.next;
    return /* Cons */{
            _0: c.content,
            _1: (function (param) {
                return aux(next, param);
              })
          };
  };
  const partial_arg = q.first;
  return function (param) {
    return aux(partial_arg, param);
  };
}

function add_seq(q, i) {
  Stdlib__Seq.iter((function (x) {
          add(x, q);
        }), i);
}

function of_seq(g) {
  const q = {
    length: 0,
    first: /* Nil */0,
    last: /* Nil */0
  };
  add_seq(q, g);
  return q;
}

const push = add;

const pop = take;

const top = peek;

export {
  Empty ,
  create ,
  add ,
  push ,
  take ,
  take_opt ,
  pop ,
  peek ,
  peek_opt ,
  top ,
  clear ,
  copy ,
  is_empty ,
  length ,
  iter ,
  fold ,
  transfer ,
  to_seq ,
  add_seq ,
  of_seq ,
}
/* No side effect */
